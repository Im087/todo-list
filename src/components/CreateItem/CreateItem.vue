<template>
  <v-container class="d-flex">
    <v-text-field
      placeholder="What are you gonna do?"
      v-model="toDoItem.toDoContent"
    >
    </v-text-field>
    <v-btn color="info" size="x-large" @click="addToDoItem">Create</v-btn>
  </v-container>
</template>

<script lang="ts">
import { useStore } from "@/store";
import { defineComponent, ref, type Ref } from "vue";

import type { ToDoItem } from '../../interfaces';

export default defineComponent ({
  name: "CreateItem",
  setup() {
    const store = useStore();
    const toDoItem: Ref<ToDoItem> = ref({
      itemStatus: false,
      timeStamp: '',
      toDoContent: '',
    });

    const addToDoItem = () => {
      toDoItem.value.timeStamp = new Date().toISOString();
      let item = Object.assign({}, toDoItem.value)
      toDoItem.value.toDoContent = '';
      store.dispatch('addToDoItem', item);
    };

    return {
      toDoItem,
      addToDoItem
    };
  }
});
</script>

<style lang="scss" scoped>

</style>