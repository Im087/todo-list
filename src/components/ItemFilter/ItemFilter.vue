<template>
  <v-container>
    <v-radio-group
      v-model="isChecked"
      hide-details inline
      @change="changeItemFilter()"
      data-test-id="item-filter__radio-group"
    >
      <v-radio
        label="All"
        color="info"
        value="all"
      >
      </v-radio>
      <v-radio
        label="Done"
        color="info"
        value="done"
      >
      </v-radio>
      <v-radio
        label="Undone"
        color="info"
        value="undone"
      >
      </v-radio>
    </v-radio-group>
  </v-container>
</template>

<script lang="ts">
import { defineComponent, ref, toRefs } from "vue";

export default defineComponent ({
  name: 'ItemFilter',
  props: {
    itemFilter: {
      type: String,
      required: true
    }
  },
  setup(props, {emit}) {
    const itemFilter = toRefs(props).itemFilter;
    const isChecked = ref(itemFilter.value);
    
    const changeItemFilter = () => {
      emit('update:itemFilter', isChecked.value);
    };

    return {
      isChecked,
      changeItemFilter
    };
  }
});
</script>

<style>

</style>