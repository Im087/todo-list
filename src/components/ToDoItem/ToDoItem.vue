<template>
  <v-container class="d-flex">
    <v-checkbox
      class="flex-grow-0"
      :label="itemStatus ? 'Done' : 'Undone'"
      :value="itemStatus ? 'Done' : 'Undone'"
      color="info"
    >
    </v-checkbox>
    <v-textarea
      class="flex-grow-1"
      rows="1"
      bg-color="transparent"
      :value="toDoContent"
      clearable
      auto-grow
      no-resize
    >
    </v-textarea>
    <v-btn color="warning" size="large">Edit</v-btn>
    <v-btn color="error" size="large" @click="deleteToDoItem(timeStamp)">Delete</v-btn>
  </v-container>
</template>

<script lang="ts">
import { defineComponent, toRefs } from "vue";
import { useStore } from "@/store";

export default defineComponent ({
  name: "ToDoItem",
  props: {
    toDoItem: {
      type: Object,
      required: true,
      deep: true
    }
  },
  setup(props) {
    const store = useStore();
    const toDoItem = toRefs(props).toDoItem;
    const itemStatus = toDoItem.value.itemStatus;
    const timeStamp = toDoItem.value.timeStamp;
    const toDoContent = toDoItem.value.toDoContent;

    const deleteToDoItem = (timeStamp: string) => {
      store.dispatch('deleteToDoItem', timeStamp);
    }

    return {
      itemStatus,
      timeStamp,
      toDoContent,
      deleteToDoItem
    };
  }
});
</script>

<style lang="scss" scoped>

</style>