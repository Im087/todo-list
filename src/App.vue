<template>
  <!-- <v-app>
    <v-container>
      <h1>To Do List</h1>
    </v-container>
    <v-main>
      <CreateItem></CreateItem>
      <ItemFilter v-model:itemFilter="itemFilter"></ItemFilter>
      <ToDoItemVue
        v-for="toDoItem in filteredArray"
        :key="toDoItem.timeStamp"
        :toDoItem="toDoItem"
      >
      </ToDoItemVue>
    </v-main>
  </v-app> -->
  <v-container>
    <v-container>
      <h1>To Do List</h1>
    </v-container>
    <v-container>
      <CreateItem></CreateItem>
      <ItemFilter v-model:itemFilter="itemFilter"></ItemFilter>
      <ToDoItemVue
        v-for="toDoItem in filteredArray"
        :key="toDoItem.timeStamp"
        :toDoItem="toDoItem"
      >
      </ToDoItemVue>
    </v-container>
  </v-container>
</template>

<script lang="ts">
import { computed, defineComponent, ref, type ComputedRef } from 'vue';

import CreateItem from './components/CreateItem/CreateItem.vue';
import ItemFilter from './components/ItemFilter/ItemFilter.vue';
import ToDoItemVue from './components/ToDoItem/ToDoItem.vue';

import type { ToDoItem } from './interfaces';
import store from './store';

export default defineComponent({
  name: 'App',
  components: {
    CreateItem,
    ItemFilter,
    ToDoItemVue,
  },
  setup() {
    const toDoArray: ComputedRef<ToDoItem[]> = computed(() => store.getters.getToDoArray);
    const itemFilter = ref('all');

    const filteredArray: ComputedRef<ToDoItem[]> = computed(() => {
      let isChecked = itemFilter.value;
      if(isChecked === 'done') {
        return toDoArray.value.filter(item => item.itemStatus === true);
      } else if(isChecked === 'undone') {
        return toDoArray.value.filter(item => item.itemStatus === false);
      } else {
        return toDoArray.value;
      }
    });

    return {
      toDoArray,
      itemFilter,
      filteredArray
    };
  }
});
</script>
