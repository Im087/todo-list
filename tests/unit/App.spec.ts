import { describe, test, expect, afterEach, vi, beforeEach } from 'vitest';
import App from '../../src/App.vue';
import CreateItem from '../../src/components/CreateItem/CreateItem.vue';
import ItemFilter from '../../src/components/ItemFilter/ItemFilter.vue';
import ToDoItem from '../../src/components/ToDoItem/ToDoItem.vue';
import { createVuetify } from "vuetify";
import * as components from "vuetify/components";
import * as directives from "vuetify/directives";
import { mount, VueWrapper } from '@vue/test-utils';
import { createStore } from 'vuex';

describe('App.vue', () => {
  const vuetify = createVuetify({components, directives});
  const undoneItem = {
    itemStatus: false,
    isDisabled: true,
    timeStamp: new Date().toISOString(),
    toDoContent: 'todo1',
  };
  const doneItem = {
    itemStatus: true,
    isDisabled: true,
    timeStamp: new Date().toISOString(),
    toDoContent: 'todo2',
  };
  const $store = {
    state: {
      toDoArray: [undoneItem, doneItem]
    },
    getters: {
      // getToDoArray: (state) => vi.fn().mockReturnValue(state.toDoArray)
      // getToDoArray: (state) => state.toDoArray
      getToDoArray: [undoneItem, doneItem]
    }
  };
  const store = createStore({
    state: {
      toDoArray: [undoneItem, doneItem]
    },
    getters: {
      // getToDoArray: (state) => vi.fn().mockReturnValue(state.toDoArray)
      getToDoArray: (state) => state.toDoArray
    }
  });

  describe('When mounted', () => {
    let wrapper: VueWrapper<any>;

    beforeEach(() => {
      wrapper = mount(App, {
        global: {
          plugins: [vuetify, store],
          // mocks: {
          //   $store
          // }
        }
      });
    });

    afterEach(() => {
      wrapper.unmount();
    });

    test('render child components', () => {
      expect(wrapper.findComponent(CreateItem).exists()).toBeTruthy();
      expect(wrapper.findComponent(ItemFilter).exists()).toBeTruthy();
    });
    // test('get vuex state', () => {
    //   expect(wrapper.vm.toDoArray).toEqual(store.getters.getToDoArray);
    // });
  });

  describe('When clicking on CREATE button', () => {
    let wrapper: VueWrapper<any>;

    beforeEach(async () => {
      wrapper = mount(App, {
        global: {
          plugins: [vuetify]
        }
      });

      await wrapper.findComponent(CreateItem).find('button').trigger('click');
    });

    afterEach(() => {
      wrapper.unmount();
    });


    test('a new item will be added to toDoArray/filteredArray', () => {
      expect(wrapper.vm.filteredArray).toHaveLength(1);
    });

    test('a ToDoItem component will be rendered', () => {
      expect(wrapper.findComponent(ToDoItem).exists()).toBeTruthy();
    });

  });

  describe('When changing filter', async () => {
    const wrapper = mount(App, {
      global: {
        plugins: [vuetify],
        // mocks: {
        //   $store
        // }
      },
      attachTo: document.body
    });

    await wrapper.findComponent(ItemFilter).findAll('input')[1].trigger('click');

    test('itemFilter will change', () => {
      expect(wrapper.vm.itemFilter).toEqual('done');
    });

    wrapper.unmount();
  });

  // describe('When itemFilter changes', async () => {
  //   const wrapper = mount(App, {
  //     global: {
  //       plugins: [vuetify],
  //       mocks: {
  //         $store
  //       }
  //     },
  //     attachTo: document.body
  //   });

  //   wrapper.vm.itemFilter = 'done';

  //   test('filteredArray will change', () => {
  //     expect(wrapper.vm.filteredArray).toEqual([doneItem]);
  //   });
  // });

  describe('snapshot', () => {
    const wrapper = mount(App, {
      global: {
        plugins: [vuetify]
      }
    });

    test('is OK', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    wrapper.unmount();
  });
});