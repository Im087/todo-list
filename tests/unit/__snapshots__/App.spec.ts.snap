// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App.vue > snapshot > is OK 1`] = `null`;
